src/wallet/api.rs

```diff
@@ -1004,40 +1019,169 @@ impl Wallet {
         Ok(())
     }
 
-    /// Largerst to lowest coinselect algorithm
-    // TODO: Fix Coin Selection algorithm for Dynamic Feerate
+    /// Performs coin selection to choose UTXOs that sum to a target amount.
+    ///
+    /// Uses the rust-coinselect library to implement Bitcoin Core's coin selection algorithm.
+    /// The algorithm tries to minimize the number of inputs while accounting for:
+    /// - Transaction fees and weight
+    /// - Long-term UTXO pool management
+    /// - Change output costs
+    /// - Privacy considerations
+    ///
+    /// # Arguments
+    /// * `amount` - The target amount to select coins for
+    ///
+    /// # Returns
+    /// * `Ok(Vec<(ListUnspentResultEntry, UTXOSpendInfo)>)` - Selected UTXOs and their spend info
+    /// * `Err(WalletError)` - If coin selection fails or there are insufficient funds
+    ///
+    /// # Note
+    /// Only considers spendable UTXOs (regular coins and swap coins), filtering out:
+    /// - Fidelity bond UTXOs
+    /// - Locked UTXOs
+    /// - Unconfirmed UTXOs
     pub fn coin_select(
         &self,
         amount: Amount,
+        feerate: f64,
     ) -> Result<Vec<(ListUnspentResultEntry, UTXOSpendInfo)>, WalletError> {
         let all_utxos = self.get_all_locked_utxo()?;
 
+        // Get spendable UTXOs (regular coins and swap coins)
         let mut seed_coin_utxo = self.list_descriptor_utxo_spend_info(Some(&all_utxos))?;
         let mut swap_coin_utxo = self.list_incoming_swap_coin_utxo_spend_info(Some(&all_utxos))?;
         seed_coin_utxo.append(&mut swap_coin_utxo);
 
-        // Fetch utxos, filter out existing fidelity coins
-        let mut unspents = seed_coin_utxo
+        // Filter out existing fidelity coins
+        let unspents: Vec<_> = seed_coin_utxo
             .into_iter()
             .filter(|(_, spend_info)| !matches!(spend_info, UTXOSpendInfo::FidelityBondCoin { .. }))
-            .collect::<Vec<_>>();
+            .collect();
 
-        unspents.sort_by(|a, b| b.0.amount.cmp(&a.0.amount));
+        // Construct TxIn from unspents
+        let inputs: Vec<TxIn> = unspents
+            .iter()
+            .map(|(utxo, _)| TxIn {
+                previous_output: OutPoint {
+                    txid: utxo.txid,
+                    vout: utxo.vout,
+                },
+                script_sig: utxo.script_pub_key.clone(),
+                sequence: Sequence::MAX, // Default sequence number
+                witness: utxo
+                    .witness_script
+                    .as_ref()
+                    .map_or_else(Witness::default, |script| {
+                        let mut witness = Witness::new();
+                        witness.push(script.as_bytes());
+                        witness
+                    }),
+            })
+            .collect();
```

Comment:
This entire input you are creating just to estimate an average size of inputs. Where there only variable is the signature size. We already have almost exact estimate of witness size (+- 2 bytes) in the new spend APIs.

So you should also be able to calculate an average weight of input based on that. 

This should just be a global constant, with a note of calculation showing how you arrived at the average value.